### ISSUE TITLE
x-on:click called twice after livewire call under certain circumstances

### ISSUE THREAD

**Issue by jdssem** [livewire/livewire#763]
**Describe the bug**
`x-on:click` event called twice after livewire communication under certain circumstances.

**To Reproduce**
- laravel v7.7.2
- livewire v1.0.9
- alpinejs v2.2.1

### Test.php
```php
class Test extends Component
{
    public $name;

    public $items;

    public function mount()
    {
        $this->items = [1, 2];
    }

    public function render()
    {
        return view('livewire.test');
    }
}
```

### test.blade.php
```html
<div>
    <input type="text" wire:model="name"/>

    @foreach($items as $item)
        <div x-data="{ open:false }" wire:key="{{ $item }}" x-bind:key="{{ $item }}">
            <div> <!-- If I remove this div, it works fine. -->
                <button x-on:click="console.log('clicked');">
                    click
                </button>
            </div>
        </div>
    @endforeach
</div>
```

1. click the button and see the browser console. it works nomally.
2. put some characters in the input box. It will make a livewire call.
3. click the button again and see the browser console. it will be called twice.

**Expected behavior**
It should be called once.

**Screenshots**
![livewire](https://user-images.githubusercontent.com/978944/77432415-607cf700-6e21-11ea-83af-508ae27b7248.gif)

**Desktop (please complete the following information):**
 - chrome, safari, firefox

**Additional context**
I have originally made a replies component. It has several buttons like edit and delete in every replies. When I click delete button, it shows a delete modal. If I click 'delete' button, it calls livewire action using javascript. But If I make some livewire actions such as deleting or editing, delete action runs twice. The code above is the simplified version of that.

**Comment by vrajroham** [comment]
@jdssem Encountered same issue. On clicking the submit button, only one xhr request is sent to a server to the specified function `E.g. wire:submit.prevent="submit"` but the code in submit function runs twice. 

In my case, 2 records are created with same input and 2 different otps are being sent.

**Comment by vrajroham** [comment]
@jdssem @calebporzio @lancepioch 

**Update** : After removing id from the form attribute all works perfectly no more running the twice. @jdssem Check if this solves yours too.

**Comment by jordyvanderhaegen** [comment]
Ran into the same issue, it looks like 2 click events are present on the DOM element after the component is updated.

**Comment by nuernbergerA** [comment]
Hey @jdssem I follow your steps and all works fine, even with your alpine and livewire versions.

Have I done something wrong?

Cheers Adrian

**Comment by naro-code-lab** [comment]
I ran into the same issue.

I added **id** to the form attribute and it worked for me.

**Comment by jskorlol** [comment]
I am the same issue.

(in foreach)
When an intermediate element is deleted from the corresponding bound item, when it is newly re-rendered
Alpine.js seems to overlap the events of the missing element in the middle.

[1,2,3,4] Then, put the click event and remove the element in the middle.

If you remove 2, element 3 will trigger a 2 click event,
If you remove 2,3, element 4 will trigger a 3 click event.

However, if you delete an element upside down 4,3,2, it works fine.

**Comment by ahamilton9** [comment]
I've run into a similar issue. Same problem, just a different Blade directive. I've tried to put together another test case below. I have not tried this on a fresh install yet but will try to soon. It is very possible there is still something else in how my team's Blade layout is ordered or something else we are loading that is causing the issue if it doesn't occur on a fresh install.

I configured this component to hide the majority of the DOM until it is "enabled", since it won't always be needed by users and is a sizable chunk in our actual code. Seemed to work well until I noticed the event was being called twice when the element that contained the click event was rendered after a Livewire component update. Works fine when displayed with the initial render as shown in the example. Let me know if I can provide more info, will update if I get a chance to reproduce fresh. I'll try the other cases in this thread as well.

**Laravel:** 7.17.2
**Livewire:** 1.2.0
**Alpine:** 2.4.1
**Testing in Firefox:** 78.0.2 (64-bit), Ubuntu Linux

**Test.php**
```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class Test extends Component
{
    public $enabled = false;

    protected $listeners = [
        'enable' => 'enable',
    ];

    public function render() {
        return view('livewire.test');
    }

    public function enable() {
        $this->enabled = true;
    }
}
```

**test.blade.php**
```html
<div
    id="test"
    x-data
    @custom-event.window="console.log($event);"
>
    <button @click="livewire.emitTo('test', 'enable');">Enable</button>

    <br />

    One event bound in inspector, click calls just once:
    <button @click="$dispatch('custom-event')">Test 1</button>

    @if($enabled)
        <br />

        Two events bound in inspector, click dispatches event twice:
        <button @click="$dispatch('custom-event')">Test 2</button>
    @endif
</div>
```

**Comment by ahamilton9** [comment]
I think I've got a more specific case, so I'm making this a new comment. If:

- Livewire adds elements to the DOM AND
- That chunk of elements was added to the end of a container AND
- Elements in that chunk of elements that have Alpine event properties on them are not directly wrapped in a NEW Alpine component (x-data'ed, existing parent Alpine component doesn't seem to help)

The events on those elements will be attached twice. Still not sure if this is due to Livewire or Alpine, sadly.
 
This works for the author of this thread's case: The foreach is adding elements, that chunk is at the end of the outermost div, and the commented div is preventing the evented element from being a child of a new Alpine component, so the events get attached twice.

I'm attaching my latest test code for my if directive case. It also falls in line with the details I described above, minus one caveat I included a note about in the code. Feels a bit like a parsing issue in how Livewire adds elements or how Alpine adds it's events, but that's just an assumption. 

### Test Code
**Test.php**
```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class Test extends Component
{
    public $enabled = false;

    protected $listeners = [
        'enable' => 'enable',
    ];

    public function render() {
        return view('livewire.test');
    }

    public function enable() {
        $this->enabled = true;
    }
}
```
**test.blade.php**
```html
<div
    id="test"
    x-data
    @custom-event.window="console.log($event);"
>
    <p>
        Button exists on initialization. One event bound in inspector, click calls just once as expected (check console.log):
        <button @click="$dispatch('custom-event')">Test 1</button>
    </p>

    <p>
        Update a Livewire property to add similar buttons to the DOM:
        <button @click="livewire.emitTo('test', 'enable');">Load Other Tests</button>
    </p>

    @if($enabled)
        <p>
            This chunk was not added to the end of #test, and works as expected (one event):
            <button @click="$dispatch('custom-event')">Test 2</button>
        </p>
    @endif

    <div>This element must not contain Blade directives I guess? Removing it breaks Test 2 even though there is a div after this new chunk of elements.</div>

    <div>
        @if($enabled)
            <p x-data>
                This chunk was added to the end of it's container, but the evented elements are contained in a new Alpine component. Works fine:
                <button @click="$dispatch('custom-event')">Test 3</button>
            </p>
        @endif
    </div>

    <div>
        @if($enabled)
            <p>
                This chunk was added to the end of it's container and the evented elements are not contained in new Alpine elements. Events are attached twice:
                <button @click="$dispatch('custom-event')">Test 4</button>
            </p>
        @endif
    </div>
</div>
```

**Comment by XaBerr** [comment]
I am not sure that this will help you but I have found a temporary solution to the problem.
The problem appears when you put the value used in foreach in the x-on:click (or @click) event.
So I changed where I put that value.
From this:
```html
<button @click="killRowId = {{$test['pk_test']}}; console.log(killRowId);">
```
to this:
```html
<button @click="killRowId = $event.currentTarget.getAttribute('pk'); console.log(killRowId);" pk="{{$test["pk_test"]}}">
```

**Comment by jeffochoa** [comment]
I found the same issue while removing elements from a list.

Before:

```html
<div x-data="{ id: {{$user->id}} }">
    <button x-on:click.prevent="$dispatch('confirm-deletion', {'handler': 'inviteDeletionConfirmed', data: {{$user->id}} });>
</div>
```

After:

```html
<div x-data="{ id: {{$user->id}} }">
    <button x-on:click.prevent="$dispatch('confirm-deletion', {'handler': 'inviteDeletionConfirmed', data: id});>
</div>
```

Basically the problem is that you cannot pass the values hardcoded to the function, the workaround is passing a reference to the value instead.

Don't do:

```javascript
$dispatch('my-event, {id: 1}
```

Do

```javascript
$dispatch('my-event', {id: myVariable })
```

**Comment by maximogithub** [comment]
DO not to use directive wire:click in the button.

use <button type="submit">Submit</button>

it worked for me.

**Comment by joshhanley** [comment]
I have run into the same issue using Livewire 2.3.1 and Alpine 2.7.3. For anyone else experiencing this, this is how I got around it. 

I had the x-on:click on the div inside the loop, but that was causing the event to fire twice. Moving the click handler outside of the loop fixed the issue.

```
<div x-on:click="selectItem()">
    @foreach($results as $key => $result)
        <div
            :class="{ 'bg-blue-500' : focusIndex == {{ $key }}}"
            class="px-2"
            wire:key="{{ $inputProperty.$key }}"
            x-on:mouseenter="focusIndex = {{ $key }}"
            x-on:mouseenter="focusIndex = null"
        >
            <x-dynamic-component :component="$resultComponent" :model="$result" />
        </div>
    @endforeach
</div>
```

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by joshhanley** [comment]
REOPEN

**Comment by joshhanley** [comment]
Sorry to reopen such an old issue, but I am still experiencing this problem. I have run into it over the last couple of days.

If you need any more information, I'd be happy to reproduce this in a repo or a test!

**Comment by lpheller** [comment]
I have run into the same issue too. Posted about this yesterday in another slightly different issue, where my "solution" to the original problem produces this exact bahaviour: firing alpine clicks twice. https://github.com/livewire/livewire/issues/2046#issuecomment-733784393

**Comment by joshhanley** [comment]
@lpheller yeah it has to do with how to Alpine click handlers are registered when inside a blade conditional. See my comment further up with how I worked around it.

I've submitted a PR #1995 which adds a failing test that replicates the issue. But I don't know enough about the Alpine watchers to fix it.

**Comment by dvic** [comment]
I have run into the same issue, originally opened in #2126:

### Description

Adding children that have a click handler to a parent that has `x-data` results in the click handlers being executed twice (when clicked).

Repo demonstrating the issue: https://github.com/dvic/hello-livewire
Originally reported in: https://github.com/alpinejs/alpine/issues/756

### Exact steps to reproduce
1. Open the JS console in your browser.
2. On the home page, click on the button *first click here*.
3. A new item is added to the `ul`.
4. Click on the added `li`.
5. The console shows that the handler is called twice (printing the text `i was clicked` twice).

### Stripped-down, copy-pastable code snippets
Full demo in https://github.com/dvic/hello-livewire

**Context**
- Livewire version: 2.3
- Laravel version: 2.5
- Alpine version: 2.7.3
- Browser: Firefox, Chrome

**Comment by dvic** [comment]
This has the label `Needs Better Reproducibility`, is this still the case? In my previous comment I have provided a repo that demonstrates the issue (https://github.com/dvic/hello-livewire).

**Comment by joshhanley** [comment]
@dvic I think that labels pretty old! The issues clear now I think, but trying to understand how it all works is now the challenge so it can be fixed 😬

**Comment by antonioribeiro** [comment]
My app is using `livewire/livewire v2.3.1`. In my case I have a really simple Liveware form, it's just for the user to click a button and fire a job. Job was firing 2 (my local) and 3 (on production, of course) times, after deep investigation of Horizon config, Redis queues, and a lot of other stuff I found that

This button

``` html
<button wire:click="generate" type="button" class="button button--primary button--medium">
    Generate report
</button>
```
Was executing this method twice:

``` php
public function generate()
{
    $this->validate($this->rules);

    ReportTransactions::dispatch($this->date);
}
```

**Comment by joshhanley** [comment]
@antonioribeiro try adding a unique wire:key to each of the buttons, for example for the first button add `wire:key="generateReset"` and the second add `wire:key="generateStart"`.

What is happening, due to the blade conditional, morphdom won't know to remove the original click handler as it thinks the two buttons are the same, instead its just changing some of the properties but keeping the same button in the DOM. Adding wire key tells it explicitly that they are different buttons, and to remove the original and create a new one.

Hope this helps!

**Comment by bramcou** [comment]
For anyone also facing this issue and for whom the wire:key="{{ $loop->index }}" solution doesn't work, another workaround is to put a boolean protected property in your Livewire component, for instance $run which is false by default. The function called by wire:click then has a if() condition in the beginning of the method to only run if $this->run == false, on the next line you set $this->run = true;

That way, regardless of the number of events, the method will run only once. Important to make the property protected so that it is reset to false for every Livewire roundtrip.

Regarding the issue itself: I'm running multiple loops in which wire:click attributes are placed but all those elements have their own unique (also unique from the other loops) wire:key attribute. Regardless, anywhere from 2 to 8 payloads for the wire:click are added to the XHR requestheaders...

Anyone closer to resolving the issue (instead of workarounds)?

**Comment by crazedVic** [comment]
Ok i was running into the same issue but once i figured out why it seemed pretty obvious:

helloworld.blade.php:
```
<div x-data>
    <button class="myButton" x-on:click="$wire.showButton()">ShowButton</button>
    @if($buttonVisible)
        <button class="myButton"  x-on:click="$wire.test()">RunTest</button>
    @endif
</div>
```

Helloworld.php:
```
class Helloworld extends Component
{
    public $buttonVisible = false;

    public function showButton(){
        $this->buttonVisible = true;
    }
    public function test(){
        error_log('test has been called');
    }
}
```
Due to the @if statement, initially the RunTest button is not rendered.
Once I click ShowButton, the buttonVisible property changes and the @if now rings true, showing the RunTest button.
Tapping the RunTest button will show only one call to the server in the Network tab.
However on the server the function runs twice:
```
[Wed Jan 20 12:45:22 2021] test has been called
[Wed Jan 20 12:45:22 2021] test has been called
```
changing the blade component to this fixes the issue:
```
<div>
    <div x-data>
        <button class="myButton" x-on:click="$wire.showButton()">Show Button</button>
    </div>
        @if($buttonVisible)
            <div x-data>
                <button class="myButton"  x-on:click="$wire.test()">Run Test</button>
            </div>
        @endif
    </div>
</div>
```

**Comment by ludgerey** [comment]
As @joshhanley mentioned before, adding a unique key helps. But if you use this in a loop, make sure every event binding has its own key. This worked for me:

```blade
<button
    wire:key="event-bind-{{ \Illuminate\Support\Str::random() }}"
    x-on:click.stop="delete({{ $model->id }})"
>
    Delete
</button>
```

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by austenc** [comment]
Hey folks, to fix this please upgrade Alpine to the latest version (`2.8.1+`)

**Comment by dvic** [comment]
@austenc thanks! I can confirm it fixes also #2126 (which was the same problem but another github issue)

**Comment by snapey** [comment]
I had this issue and was _not_ using alpine.

Looking in the dom at the event listeners on the button element, an additional click event was being added every time the dom was refreshed.  Resolved it by keeping the buttons in the dom all the time, but adding or removing css hidden class only.

**Comment by adddz** [comment]
> Ok i was running into the same issue but once i figured out why it seemed pretty obvious:
> 
> helloworld.blade.php:
> 
> ```
> <div x-data>
>     <button class="myButton" x-on:click="$wire.showButton()">ShowButton</button>
>     @if($buttonVisible)
>         <button class="myButton"  x-on:click="$wire.test()">RunTest</button>
>     @endif
> </div>
> ```
> 
> Helloworld.php:
> 
> ```
> class Helloworld extends Component
> {
>     public $buttonVisible = false;
> 
>     public function showButton(){
>         $this->buttonVisible = true;
>     }
>     public function test(){
>         error_log('test has been called');
>     }
> }
> ```
> 
> Due to the @if statement, initially the RunTest button is not rendered.
> Once I click ShowButton, the buttonVisible property changes and the @if now rings true, showing the RunTest button.
> Tapping the RunTest button will show only one call to the server in the Network tab.
> However on the server the function runs twice:
> 
> ```
> [Wed Jan 20 12:45:22 2021] test has been called
> [Wed Jan 20 12:45:22 2021] test has been called
> ```
> 
> changing the blade component to this fixes the issue:
> 
> ```
> <div>
>     <div x-data>
>         <button class="myButton" x-on:click="$wire.showButton()">Show Button</button>
>     </div>
>         @if($buttonVisible)
>             <div x-data>
>                 <button class="myButton"  x-on:click="$wire.test()">Run Test</button>
>             </div>
>         @endif
>     </div>
> </div>
> ```
> 
> It would be nice to just have alpine trigger an exception instead of displaying this weird and unexpected behaviour.

Thank you so much, this really helped. I was going nuts with this issue.
With Alpine 2, this was working fine but for some reasons, not with Alpine 3.

Anyone can explain why this wasn't an issue in Alpine 2?

**Comment by alemhulu** [comment]
<button type="button" class="hidden"  id="primaryButton" onclick="alert('Hello!')" > </button>
<button type="button"  id="secondaryButton"
  onclick="document.getElementById('primaryButton').click()" > </button>

give for the first button id and on  the second get the element by id of the first button  and add click function by as the shown above.

**Comment by alemhulu** [comment]
<input type="button" id="primaryButton" onclick="alert('Hello!')" />
<input type="button"
  id="secondaryButton"
  onclick="document.getElementById('primaryButton').click()" />

**Comment by alemhulu** [comment]
<input type="button" id="primaryButton" onclick="alert('Hello!')" />
<input type="button"
  id="secondaryButton"
  onclick="document.getElementById('primaryButton').click()" />

**Comment by gmartinez** [comment]
Probably not the same for folks above, though while going nuts on same issue (lw method being called twice) I ended up here.

I eventually found that I was doing an _onclick="Livewire.emit(methodName)"_ while under the very same lw component's blade template.
See, this was an op that previously was running from outside the component, but copy & paste html into it carried the emit thing.

Anyway, just in case someone out there as stupid as me.
